gax-grpc/src/main/java/com/google/api/gax/grpc/ClientContext.java

```diff
@@ -27,43 +27,105 @@
  * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
-package com.google.api.gax.grpc;
+package com.google.api.gax.rpc;
 
+import com.google.api.core.ApiClock;
 import com.google.api.core.BetaApi;
+import com.google.api.core.CurrentMillisClock;
+import com.google.api.gax.core.ExecutorProvider;
 import com.google.auth.Credentials;
 import com.google.auto.value.AutoValue;
-import io.grpc.Channel;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
 import javax.annotation.Nullable;
 
 /**
- * Encapsulates client state, including channel, executor, and credentials.
+ * Encapsulates client state, including executor, credentials, and transport context.
  *
  * <p>Unlike {@link ClientSettings} which allows users to configure the client, {@code
  * ClientContext} is intended to be used in generated code. Most users will not need to use it.
  */
 @BetaApi
 @AutoValue
 public abstract class ClientContext {
-  public abstract Channel getChannel();
+
+  /**
+   * The objects that need to be closed in order to clean up the resources created in the process of
+   * creating this ClientContext. This will include the closeables from the transport context.
+   */
+  public abstract Collection<AutoCloseable> getCloseables();
```

Comment:
I think this should be a list. We need to close things out in opposite order than how they are created.